crates/via-serve-static/src/stream_file.rs

```diff
@@ -117,53 +116,22 @@ fn stream_file_blocking(sender: mpsc::Sender<ReadChunkResult>, path: PathBuf, ti
 impl StreamFile {
     pub fn new(path: PathBuf, timeout: u64) -> Self {
         let (sender, receiver) = mpsc::channel(CHANNEL_CAPACITY);
-        let mut results = Vec::new();
-
-        results.reserve_exact(CHANNEL_CAPACITY);
 
         // Spawn a blocking task to read the file in chunks.
         task::spawn_blocking(move || stream_file_blocking(sender, path, timeout));
 
-        Self { receiver, results }
+        Self { receiver }
     }
 }
 
 impl Stream for StreamFile {
     type Item = Result<Vec<u8>>;
 
     fn poll_next(self: Pin<&mut Self>, context: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        // Get a mutable reference to the pinned data.
         let this = self.get_mut();
 
-        loop {
-            // Check whether or not we have results in `self.results`. If so,
-            // pop the last result and return it.
-            if let Some(result) = this.results.pop() {
-                return Poll::Ready(Some(result.map_err(Error::from_io_error)));
-            }
-
-            // Create an intermediate buffer to hold received chunks. While it's
-            // possible to reverse the order of the chunks in the results buffer,
-            // it's probably safer to extend `self.results` with a reversed
-            // iterator.
-            let mut buf = Vec::new();
-
-            buf.reserve_exact(CHANNEL_CAPACITY);
-
-            // Poll the receiver for new data.
-            match this
-                .receiver
-                .poll_recv_many(context, &mut buf, CHANNEL_CAPACITY)
-            {
-                // No data yet.
-                Poll::Pending => return Poll::Pending,
-                // No more data.
-                Poll::Ready(0) => return Poll::Ready(None),
-                // Extend `self.results` with the elements in `buf`.
-                Poll::Ready(_) => {
-                    this.results.extend(buf.into_iter().rev());
-                }
-            }
-        }
+        this.receiver
```

Comment:
The performance of the implementation using `poll_recv_many` is roughly the same as that using `poll_recv`. Given that the `ReadChunkResult` messages are already buffered behind a semaphore in the channel, the Stream implementation using `poll_recv` is likely more secure and correct.